### ISSUE TITLE
[psionic][perf] Mirror llama.cpp GPT-OSS graph and CUDA fusion architecture

### ISSUE THREAD

**Issue by AtlantisPleb** [OpenAgentsInc/openagents#3249]
## Problem

Psionic is now materially faster than its original GPT-OSS baseline, but the latest measured benchmark still shows a large remaining gap on the exact same host/model/HTTP flow:

- Psionic: `35.26 tok/s`
- `llama.cpp`: `167.27 tok/s`

The latest checkpoint also falsified one of the earlier bottleneck theories. After adding a CUDA argmax fast path, Psionic now reads back only `148` device-to-host bytes across the timed request, yet throughput barely moved. That means the dominant remaining gap is no longer logits readback. The real gap is that Psionic still does not mirror `llama.cpp`'s OpenAI-MoE graph, fusion, and dispatch architecture.

### Psionic current hot path

- `crates/psionic/psionic-serve/src/gpt_oss.rs`
  - Rust-owned decode-step plan encoded every token
  - explicit per-layer sequencing for RMSNorm -> QKV -> attention -> residual -> router -> MoE -> output
  - recent CUDA argmax fast path and residual-copy cleanup already landed
- `crates/psionic/psionic-backend-cuda/src/kernels/quantized_matvec.cu`
  - partial `llama.cpp`-style MMVQ row kernel now exists for `Q8_0` / `MXFP4`
  - still isolated from the larger graph/fusion policy

### `llama.cpp` reference path to mirror

- `~/code/llama.cpp/src/models/openai-moe-iswa.cpp`
  - exact GPT-OSS / OpenAI-MoE graph order
- `~/code/llama.cpp/src/llama-graph.cpp`
  - `build_attn(...)`
  - `build_moe_ffn(...)`
  - top-k MoE weight and expert graph construction
- `~/code/llama.cpp/src/llama-context.cpp`
  - graph reserve / reuse / scheduler integration
- `~/code/llama.cpp/ggml/src/ggml-cuda/ggml-cuda.cu`
  - subgraph fusion decisions
  - MMVQ/MMQ dispatch
  - top-k MoE fusion
  - RMSNorm fusion
  - CUDA graph execution/capture rules

## Required changes

1. Mirror the GPT-OSS graph structure, not just the math
   - make the Psionic CUDA decode/prefill representation follow the same high-level structure as `openai-moe-iswa.cpp`
   - explicitly represent the graph nodes / aliases / views needed for residuals, expert splits, and final output projection
   - stop treating the current Rust token-step encoder as the long-term architecture

2. Add a fusion/dispatch layer in Psionic that mirrors the proven `llama.cpp` decisions
   - port the relevant decision rules from `ggml-cuda.cu` for:
     - quantized MMVQ vs MMQ selection
     - `mul_mat_id` grouped-expert dispatch
     - top-k MoE fusion
     - RMSNorm fusion
     - RoPE + KV write fusion where applicable
   - the goal is not "inspired by"; it is to make Psionic choose work the same way for the GPT-OSS lane unless there is a documented intentional deviation

3. Make graph reuse a first-class runtime contract
   - add stable decode-graph identity / validity inputs
   - support CUDA graph capture or equivalent reusable execution shape for the steady-state decode token
   - surface graph-capture / reuse / rebuild evidence in perf receipts and observability

4. Keep the ownership boundary correct
   - the graph/runtime remains Psionic-owned in Rust crates
   - the CUDA kernels / launch policy can be direct ports where that is the shortest truthful path
   - do not shell out to or FFI-wrap `llama.cpp` as an execution dependency

## Direct reference code to study and port

- `~/code/llama.cpp/src/models/openai-moe-iswa.cpp`
- `~/code/llama.cpp/src/llama-graph.cpp`
- `~/code/llama.cpp/src/llama-context.cpp`
- `~/code/llama.cpp/ggml/src/ggml-cuda/ggml-cuda.cu`

## Acceptance criteria

- Psionic no longer encodes the steady-state GPT-OSS CUDA decode token as an ad-hoc Rust op list without graph/fusion semantics
- Psionic exposes graph reuse / fusion evidence for the GPT-OSS lane
- The refactor materially improves the benchmark on the exact host/model/script contract
- The resulting architecture is a direct, inspectable analogue of the `llama.cpp` GPT-OSS path rather than another isolated micro-optimization

## Dependency order

- This should run before the remaining low-level kernel-port issue and before the final parity closure issue.

**Comment by AtlantisPleb** [comment]
Checkpoint update from the current 2026-03-09 psionic GPT-OSS perf pass:\n\n- fixed MXFP4 correctness to match llama.cpp ( decode and missing  in the MXFP4 x Q8_1 path)\n- routed the real GPT-OSS CUDA lane through the Q8_1 fast path for QKV, attention output, MoE gate/up, MoE down, and final output where eligible\n- added greedy decode CUDA graph replay and fixed the request-local graph lifetime bug\n- benchmark now sits at 68.43 tok/s for psionic versus 168.72 tok/s for llama.cpp on the local RTX 4080 host\n- new concrete gap found in the exact benchmark flow: llama.cpp serves the timed request from a live prompt cache (), while psionic still rebuilds a fresh CUDA KV mirror from host-owned prefix cache state on each request\n\nTargeted verification passed with 
running 30 tests
test tests::architecture_mapping_handles_known_compute_capabilities ... ok
test tests::cuda_dense_surface_is_documented ... ok
test tests::cuda_plan_validation_rejects_unsupported_ops ... ok
test tests::mig_parser_skips_disabled_and_na_values ... ok
test tests::risk_profile_marks_display_and_mig_devices_as_elevated ... ok
test tests::recovery_profile_preserves_gpu_reset_when_driver_reports_none ... ok
test tests::inventory_row_parses_into_expected_descriptor_inputs ... ok
test tests::cuda_health_is_degraded_when_a_discovered_device_is_elevated ... ok
test tests::cuda_backend_runtime_resources_are_explicit_when_available ... ok
test tests::cuda_backend_rejects_non_cuda_tensor_specs_when_available ... ok
test tests::cuda_backend_fallback_selection_reports_explicit_cpu_fallback_when_unavailable ... ok
test tests::cuda_submission_executes_mxfp4_moe_down_aggregate_q8_1_when_available ... ok
test tests::cuda_backend_allocates_and_submits_copy_when_available ... ok
test tests::cuda_submission_executes_mxfp4_quantized_matvec_q8_1_fast_path_when_available ... ok
test tests::cuda_submission_executes_add_residual_rms_norm_when_available ... ok
test tests::cuda_backend_executes_mxfp4_quantized_matvec_when_available ... ok
test tests::cuda_submission_executes_q8_0_quantized_matvec_q8_1_fast_path_when_available ... ok
test tests::cuda_backend_refuses_allocation_when_unavailable ... ok
test tests::cuda_submission_executes_argmax_when_available ... ok
test tests::cuda_backend_executes_q8_0_quantized_matvec_when_available ... ok
test tests::cuda_submission_executes_argmax_for_wide_rows_when_available ... ok
test tests::cuda_submission_executes_q8_0_quantized_matvec_q8_1_fast_path_for_multi_block_rows_when_available ... ok
test tests::cuda_submission_executes_mxfp4_quantized_matvec_q8_1_fast_path_for_multi_block_rows_when_available ... ok
test tests::cuda_submission_executes_mxfp4_moe_gate_up_swiglu_q8_1_when_available ... ok
test tests::cuda_submission_fused_attention_matches_separate_rope_attention_and_cache_path_when_available ... ok
test tests::cuda_backend_executes_multiple_matmuls_and_adds_when_available ... ok
test tests::cuda_backend_executes_dense_surface_when_available ... ok
test tests::cuda_backend_report_is_self_consistent ... ok
test tests::cuda_submission_executes_f16_rhs_matmul_when_available ... ok
test tests::cuda_backend_selection_reports_ready_or_degraded_cuda_when_available ... ok

test result: ok. 30 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.48s


running 63 tests
test conformance::tests::generate_case_builder_uses_real_qwen2_fixture ... ok
test conformance::tests::conformance_suite_accepts_matching_prompt_render_candidate ... ok
test conformance::tests::local_ollama_catalog_subject_lists_and_shows_models ... ok
test conformance::tests::conformance_suite_fails_on_embed_drift ... ok
test conformance::tests::local_ollama_catalog_subject_reports_missing_model ... ok
test conformance::tests::compare_show_ignores_candidate_only_psionic_facts ... ok
test conformance::tests::performance_gate_accepts_candidate_within_generation_budget ... ok
test conformance::tests::performance_gate_refuses_cutover_when_metrics_are_missing ... ok
test conformance::tests::local_ollama_catalog_subject_show_reports_adapter_policy_facts ... ok
test conformance::tests::report_writer_emits_pretty_json ... ok
test conformance::tests::show_observation_from_descriptor_reports_native_boundary_facts ... ok
test gpt_oss::tests::cuda_step_plan_digest_includes_decode_graph_signature ... ok
test gpt_oss::tests::gpt_oss_decode_graph_matches_llama_cpp_high_level_order ... ok
test gpt_oss::tests::packed_expert_projection_bytes_interleave_each_expert_group ... ok
test gpt_oss::tests::packed_projection_bytes_preserve_projection_order ... ok
test gpt_oss::tests::q8_1_mmvq_helper_accepts_gpt_oss_quantization_modes ... ok
test gpt_oss::tests::quantized_matrix_transpose_decode_preserves_row_values ... ok
test gpt_oss::tests::split_projection_outputs_respects_segment_rows ... ok
test openai_http::tests::chat_messages_map_to_prompt_messages ... ok
test openai_http::tests::ensure_harmony_stop_sequences_is_idempotent ... ok
test openai_http::tests::final_assistant_content_prefers_final_channel ... ok
test openai_http::tests::generation_options_force_harmony_stop_sequences ... ok
test tests::cpu_reference_text_generation_refuses_prompt_context_overflow ... ok
test tests::cpu_reference_text_generation_rejects_unknown_session ... ok
test tests::generation_sessions_reject_descriptor_drift_even_when_model_id_matches ... ok
test tests::generation_sessions_isolate_and_reset_kv_cache ... ok
test tests::generation_request_json_is_stable ... ok
test tests::generation_sampling_options_round_trip ... ok
test tests::embedding_output_rejects_non_finite_values ... ok
test tests::embedding_request_json_is_stable ... ok
test tests::generation_response_round_trips ... ok
test tests::model_registry_records_lifecycle_transitions ... ok
test tests::embedding_response_round_trips ... ok
test tests::cpu_reference_text_generation_can_truncate_oldest_prompt_tokens ... ok
test tests::golden_prompt_render_cases_are_reusable_from_serve ... ok
test tests::golden_prompt_window_cases_are_reusable_from_serve ... ok
test tests::model_registry_refuses_candidate_when_host_budget_would_be_exceeded ... ok
test tests::model_registry_tracks_active_generation_models ... ok
test tests::cpu_reference_generation_stream_disconnect_returns_terminal ... ok
test tests::model_registry_tracks_keepalive_order_and_idle_expiry ... ok
test tests::model_registry_zero_keepalive_unloads_when_idle ... ok
test tests::golden_fixture_corpus_is_visible_from_serve ... ok
test tests::paged_kv_cache_tracks_growth_refill_and_refusal ... ok
test tests::seeded_sampling_is_replayable ... ok
test tests::model_registry_can_evict_oldest_idle_model_to_admit_new_candidate ... ok
test tests::penalties_shift_token_selection ... ok
test tests::cpu_reference_generation_stream_emits_chunks_then_terminal ... ok
test tests::cpu_reference_observability_reports_sessions_memory_and_transitions ... ok
test tests::generation_stream_returns_error_before_first_chunk_when_prompt_is_empty ... ok
test tests::cpu_reference_generation_stream_cancellation_discards_uncommitted_session_output ... ok
test tests::cpu_reference_text_generation_session_budget_counts_existing_context ... ok
test tests::shared_prefix_store_reports_hit_miss_and_rebuilt ... ok
test tests::cpu_reference_text_generation_is_deterministic ... ok
test tests::smoke_embeddings_service_is_deterministic ... ok
test tests::cpu_reference_text_generation_updates_loaded_model_residency ... ok
test tests::cpu_reference_text_generation_truncates_stop_sequences ... ok
test conformance::tests::ollama_http_subject_normalizes_live_http_responses ... ok
test tests::generation_stream_reports_runtime_failure_after_stream_start ... ok
test tests::cpu_reference_text_generation_reports_cold_then_warm_provenance ... ok
test tests::cpu_reference_text_generation_reuses_and_resets_sessions ... ok
test tests::cpu_reference_text_generation_reports_prefix_hits_and_bypasses ... ok
test tests::cpu_reference_text_generation_replays_seeded_sampling_options ... ok
test tests::psionic_local_runtime_forwards_catalog_lifecycle_generation_and_embeddings ... ok

test result: ok. 63 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 1 test
test cuda_model_backed_embeddings_match_cpu_baseline_within_tolerance_or_report_explicit_fallback ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.17s


running 1 test
test cuda_model_backed_embeddings_flow_returns_response_capability_and_receipt_or_explicit_unavailability ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.20s


running 1 test
test metal_model_backed_embeddings_parity_reports_explicit_offline_state ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 1 test
test metal_model_backed_embeddings_flow_returns_response_capability_and_receipt_or_explicit_unavailability ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 1 test
test metal_model_backed_text_generation_returns_response_capability_and_receipt_or_explicit_unavailability ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 1 test
test metal_text_generation_parity_reports_explicit_offline_state ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 5 tests
test model_backed_embeddings_service_reports_missing_artifact ... ok
test model_backed_embeddings_empty_batch_returns_empty_success ... ok
test model_backed_embeddings_reject_reference_descriptor_without_fallback ... ok
test model_backed_embeddings_reject_wrong_product ... ok
test model_backed_embeddings_flow_returns_response_capability_and_receipt ... ok

test result: ok. 5 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 4 tests
test model_backed_text_generation_service_reports_missing_artifact ... ok
test model_backed_text_generation_rejects_reference_descriptor_without_fallback ... ok
test model_backed_text_generation_rejects_unknown_session ... ok
test model_backed_text_generation_flow_returns_response_capability_and_receipt ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 2 tests
test smoke_embeddings_empty_batch_returns_empty_success ... ok
test smoke_embeddings_flow_returns_response_capability_and_receipt ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 1 test
test cpu_reference_text_generation_flow_returns_response_and_receipt ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s and . The audit and roadmap are updated to reflect this checkpoint and the next work now biases toward backend-resident prompt cache reuse plus deeper llama.cpp graph/MMVQ/flash-attention parity.
